### ISSUE TITLE
WarnUnmatchedDirective false positives

### ISSUE THREAD

**Issue by dnephin** [alecthomas/gometalinter#422]
I've encountered a few cases where the "nolint" check fails:
* when `//nolint: unconvert` is in a file with a build tag, and that file isn't being included in the build (not sure if this is all linters, but I hit it with unconvert)
* when `//nolint: unconvert` is used in a file that only needs to be whitelisted for some platforms (ex: convert  to int64() is only necessary for 32bit). I'm not sure if this one can be fixed.
* when other linters fail, a nolint is sometimes reported as being unnecessary, after other failures are fixed the `nolint` failure goes away.

**Comment by alecthomas** [comment]
Okay interesting. Any thoughts on ways to fix these? Specifically in the context of the new v3-unstable API?

**Comment by dnephin** [comment]
Still thinking about solutions, I just wanted to record the problem while it was fresh.

I think the new API would fix the first issue. Instead of using `pathsToFileGlobs()` to expand directories the `directiveParser` could receive the list of files from the `loader.Program` that would be passed to the linter, which should be only the files with the correct tag. Although that might require a separate `directiveParser` per linter (at the very least a separate list of `ignoredRange` per linter). We could still cache the AST of each file globally in that case.

The third issue is probably a bug, I need to try to reproduce it with a test case, because I'm not sure.

The second issue I'm hoping is only relevant to `unconvert`. I'm not really sure how to solve that one. One idea is that `nolint` could accept a tag (`// nolint(GOARC=[386, arm,...]`) however that is a lot more complexity, for something that is hopefully a very rare case, so I don't like that option.

However, accepting tags would also fix the first issue without needing to convert everything to the new API to pick up the fix, so maybe worth considering.

**Comment by roganartu** [comment]
If it helps, I also get this with the `golint` linter:
```
db/db.go:9:2:warning: a blank import should be only in a main or test package, or have a comment justifying it (golint)
```
after adding a per-line `// nolint: golint` annotation at the end of the line:
```
db/db.go:9:48:warning: nolint directive did not match any issue (nolint)
```

**Comment by dnephin** [comment]
Ah, that is a bit of a strange case. I think what's happening is that you're actually fixing the original `golint` issue by adding the comment, so the lint error is gone, and the `nolint` directive is correctly failing. I don't think this is a false positive.

Instead of using a `nolint` I would add a comment that explains why the blank import is necessary. I don't think this is something we can fix in `gometalinter`.

**Comment by JeanMertz** [comment]
Another problem I found, maybe related to this issue, but unsure yet:

locally, when running `gometalinter` (using `go -u get`) everything was fine. Then on the CI, I got:

```
streamconfig/kafkaconfig/configmap.go:45:47:warning: nolint directive did not match any issue (nolint)
```

But then, when I removed that (`// nolint: gas`) directive, I got this locally:

```
streamconfig/kafkaconfig/configmap.go:45::warning: Errors unhandled.,LOW,HIGH (gas)
```

Both (should) be using the exact same gometalinter version, so unsure where this difference is coming from.

The relevant line is here:

https://github.com/blendle/go-streamprocessor/blob/7d275253226a4af1aaf2b49c2fa624491572df54/streamconfig/kafkaconfig/configmap.go#L45

**Comment by fho** [comment]
We have a similiar issue, we marked some lines that trigger megacheck warnings with 
`//nolint: megacheck`.

From now and then Gometalinter warns about `nolint directive did not match any issue (nolint)` in our CI runs.
When we rerun gometalinter on the same commit (no files changed) in CI, the warning is gone.
I couldn't reproduce it on my local workstation yet. I have no clue what makes the warnings popup now and then.

**Comment by lkysow** [comment]
I also have this error occur frequently when running on CircleCI. Re-running the job fixes it.

Within the last week:
* https://circleci.com/gh/runatlantis/atlantis/1182
* https://circleci.com/gh/runatlantis/atlantis/1173
* https://circleci.com/gh/runatlantis/atlantis/1168
